[WIP] cmd: experiment (#1926)
iterative/dvc.org: content/docs/command-reference/exp/index.md

> # exp[eriments]

Reviewer: Should we explain in the description that both forms are accepted? Or just call it the official one here (and in the nav) and then create another nav entry for whichever one is an alias? Like https://dvc.org/doc/command-reference/doctor

Reviewer: Not sure about this. Thoughts @shcheklein ?

Reviewer 2: we def should do something about this (title should be clean). I think the thing we did with doctor is an overkill to support and write. I would just mention somewhere in the doc that there is an alias.

but we'll need to check if `dvc exp --help` and `dvc exeriments --help` print the right link, etc

Reviewer: > we'll need to check if dvc exp --help and dvc exeriments --help print the right link

They both print the same `dvc experiments` help output and link to https://man.dvc.org/experiments so I guess let's stick to the full name and just mention the short alias in the doc.

> A set of commands to generate and compare local _experiments_:

Reviewer: How about
```suggestion
A set of commands to generate and manage <abbr>experiments</abbr>:
```
? (which would also need a glossary entry in content\docs\user-guide\basic-concepts)

Reviewer: p.s. The keyword/term here being "experiment management".

Reviewer: BUT this brings up another issue which is: how do we clearly distinguish between regular experiments (e.g. in https://dvc.org/doc/start/experiments) and the special experiments enabled by this feature? Do we need to come up with terminology like "DVC experiment" and use it consistently (along with the tooltip I suggested above)? Cc @shcheklein

Author: The terminology is definitely an issue, we could maybe use the "ephemeral experiment vs formal/committed experiment" wording that was suggested way back in the original discussion.

Reviewer: Briefly discussed this with Ivan and it turns out there may not be a need to disambiguate: all experiments are experiments, it's just that some are committed. It's more a distinction in the way they are managed, but we can leave that explanation to an "Experiment Management" guide or use case that we should write (not in this PR of course).

[show](/doc/command-reference/exp/show),
[checkout](/doc/command-reference/exp/show),
[diff](/doc/command-reference/exp/diff), [run](/doc/command-reference/exp/run),
[resume](/doc/command-reference/exp/resume), and
[gc](/doc/command-reference/exp/gc).

## Synopsis

```usage
usage: dvc exp[eriments] [-h] [-q | -v] {show,checkout,diff,run,resume,res,gc} ...

positional arguments:
  COMMAND
    show                Print experiments.
    checkout            Checkout experiments.
    diff                Show changes between experiments in the DVC repository.
    run                 Reproduce complete or partial experiment pipelines.
    resume (res)        Resume checkpoint experiments.
    gc                  Garbage collect unneeded experiments.
```

## Types of experiments

DVC supports two types of experiments:

> 1. Standalone experiments

Reviewer: Link to https://dvc.org/doc/start/experiments ?

Reviewer: Ah no, I see you mean something else here. But still we should link to/from that page, probably. Adding the relevant notes. WDYT @pmrowla ?

Reviewer: Or did you mean that "std exps" are the old ones (one per commit) and "checkpoints" are the ones `dvc exp` uses?

Reviewer: p.s. Oh, and if not, let's make this **Types** section part of the description, please — may not even need a header. It only exists in metrics and plots refs. to redirect users to one or the other before they read an entire desc. for the wrong one, but there's no such ambiguity here.

Author: In this case standalone is still the "new" ones, just without checkpoint outputs. In this case I'm referring to using `dvc exp` to generate lots of one-per-commit/branch experiments with modified parameters and/or code, but without needing to commit them in the main repo.

Reviewer: OK. I'm not sure we want to call checkpoints a different type of experiment. Maybe it's a feature on top of experiments? I need to play with this more to be able to tell...

Reviewer: But this should def be in the **Description** section.

2. Checkpoints

## Description

> 

Reviewer: This one will be pretty important, defining what DVC Experiments are exactly, and how DVC implements in general (more details in each subcommand).

Reviewer: So actually we can leave the deep discussion on experiment management for a future guide/use case. Let's keep a very short  cmd ref index for now — we may even get rid of all these indices soon (discussed somewhere in #1944 I think).

# exp[eriments]

A set of commands to generate and compare local _experiments_:
[show](/doc/command-reference/exp/show),
[checkout](/doc/command-reference/exp/show),
[diff](/doc/command-reference/exp/diff), [run](/doc/command-reference/exp/run),
[resume](/doc/command-reference/exp/resume), and
[gc](/doc/command-reference/exp/gc).

## Synopsis

```usage
usage: dvc exp[eriments] [-h] [-q | -v] {show,checkout,diff,run,resume,res,gc} ...

positional arguments:

>   {show,checkout,diff,run,resume,res,gc}

Reviewer: ```suggestion
  COMMAND
```
(Consistency with other cmd ref index pages, I'll commit this.)

>                         Use `dvc experiments CMD --help` to display command-specific help.

Reviewer: ```suggestion
```

# exp[eriments]

A set of commands to generate and compare local _experiments_:
[show](/doc/command-reference/exp/show),
[checkout](/doc/command-reference/exp/show),

> [diff](/doc/command-reference/exp/diff), [run](/doc/command-reference/exp/run),

Reviewer: Same, run before diff, prob. Maybe also resume before diff, even, since I think it's closely linked to run.

[resume](/doc/command-reference/exp/resume), and
[gc](/doc/command-reference/exp/gc).

## Synopsis

```usage
usage: dvc exp[eriments] [-h] [-q | -v] {show,checkout,diff,run,resume,res,gc} ...

> 
> Commands to display and compare experiments.
> Documentation: <https://man.dvc.org/experiments>

Reviewer: No need for these lines from the -h output in cmd refs 🙂 I'll remove them:
```suggestion

Commands to display and compare experiments.
Documentation: <https://man.dvc.org/experiments>
```

# experiments

> A set of commands to generate and compare local <abbr>experiments</abbr>:

Reviewer: So for now this should suffice (no complications):

```suggestion
A set of commands to generate and manage experiments:
```

> Per https://github.com/iterative/dvc.org/pull/1926#discussion_r544571995. Sorry for the back and forth!

Reviewer: I'll just commit this in fact 🙂

[apply](/doc/command-reference/exp/apply),
[branch](/doc/command-reference/exp/branch),
[diff](/doc/command-reference/exp/diff), [gc](/doc/command-reference/exp/gc),
[list](/doc/command-reference/exp/list),
[pull](/doc/command-reference/exp/pull),
[push](/doc/command-reference/exp/list),
[resume](/doc/command-reference/exp/resume),
[run](/doc/command-reference/exp/run), and
[show](/doc/command-reference/exp/show).

> Note that `experiments` is also aliased to `exp`.

Reviewer: ```suggestion
Note that `dvc experiments` is also aliased to `dvc exp` including subcommands
e.g. `dvc exp list`.
```
